### ISSUE TITLE
Introduce the Rhisis domain models

### ISSUE THREAD

**Issue by Eastrall** [Eastrall/Rhisis#441]
As the time passes, we improve ourselves and see things differently; that's how most of the projects evolve.
Rhisis is not an exception, and everyday a new idea comes out of our minds. Today, I'd like to introduce a feature that could be really usefull for the future of the projects.

## Current implementation

Actually we have entities that has components which contains the data. This approach is nice because there are no business/system logic inside these entities. Some entities share the same component structure which allows us to write less code.

## The problem

In some cases, having one component shared for N entities is not a good option ; when we need to create a property for an entity, we sometimes wonder in which component the property will be stored. 

Let's take for example the `BattleComponent`. This component has a list of the targeted entities, list of projectiles and of course the information if we are fighting or not. During the fix of #438 I had to add a `ReturnToBegin` property to a monster component which is essentially triggered during battles. So, naturally I thought: "let's add it to the `BattleComponent` !" But this property will be shared with the player entity, which also uses the `BattleComponent` and don't actually need the property.

So here, we typically have an entity design issue. Let's see about an idea of fixing this.

## The idea

The idea is to implement some sort of "domain" layer, which will contain our actual entities as "model" classes with some common logic in it. 

> This approach is similar to the DDD approach: https://en.wikipedia.org/wiki/Domain-driven_design

The idea would be to have `class`es that will represent our entities and everything they need. Of course, the goal is to create simple entities.
Let's take back the battle component example:

```csharp
public class BattleComponent
{
    public WorldEntity Target { get; set; }
    public bool IsFighting => Target != null;
    // ...
}
```
we could imagine having an extra component with the monsters related properties:
```csharp
public class MonsterBattleComponent : BattleComponent
{
    public bool ReturnToBeginPosition { get; set; }
    // ...
}
```
and the same for the player:
```csharp
public class PlayerBattleComponent : BattleComponent
{
    // ...
}
```

Then when it comes to the entities, we will simple need to add the component to the right entity:
```csharp
// Let's make the entities inherit from a Mover class that has some "shared" components and propertes like the Position, Map Id, Layer Id, etc...
public class Monster : Mover
{
    // ...
    public MonsterBattleComponent Battle { get; set; }
}

public class Player : Mover
{
    // ...
    public PlayerBattleComponent Battle { get; set; }
}
```

---

I believe this could be an interesting approach to implement in the project. This could allow us to write better and cleaner code in the future.

Any thoughts?

**Comment by Eastrall** [comment]
Next steps in the reflexion of changing the way we use our entities, we need to define how they will look like according to the DDD idea exposed above.

## Base entities

Every entity in the world server, can be represented a an `object`. We can't use the `Object` class, because it will conflict with the `System.Object` class from the .NET Framework. Let's call it `WorldObject`.

A `WorldObject` can be a player, monster, dropped item, npc, or even static objects. (Houses, rocks, trees, etc...)

> ℹ️ Note that I'm including static objects in the process because they will be used in the futur to detect entity collision and prevent server hacks. 😉 

Thus, a `WorldEntity` will have the following properties:
| Name | Type | Description |
|-------|-------|-------------|
| Id | `int` | Every object among a world server will have a unique id. |
| Type | `WorldEntityType` | The world object type. (might be renamed) |
| ModleId | `int` | The 3D model id used by the object. |
| MapId | `int` | The map id where the object currently is. |
| LayerId | `int` | The current map layer id. |
| Position | `Point3D` (previously called `Vector3`) | Identifies the world object position on a map/layer. |
| Angle | `float` | Object orientation angle. |
| State | `ObjectStateType` | Defines the object state, like standing, running, jumping, etc... |
| MotionFlags | `StateFlags` | Defines the object motion, like sit, walk, etc... |
| StateMode | `StateMode` | Defines the object state mode, like normal, PK or PVP. (I believe this could be only for players thought.) |

## Moving entities

Now, that we have defined what could look like the basic world objects, we need to define how movable objects will look like. If we follow the official files, the moving objects are defined with the `CMover` class. Let's then call them `Mover`s in Rhisis.

A `Mover` object, is an object that can move from a point A to a point B at a given speed. So for now, we could consider the following properties:

> ℹ️ `Mover` inherits from the `WorldObject` properties.

| Name | Type | Description |
|-------|-------|-------------|
| Destination | `Point3D` | The destination position of the mover. |
| Speed | `float` | The mover moving base speed. |
| SpeedFactor | `float` | The mover moving speed factor. |

> This section isn't completed yet. More properties should added soon.

---

**Comment by Eastrall** [comment]
Implemented in #472
